library/cgroups.go

```diff
@@ -0,0 +1,105 @@
+package library
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/fs"
+	"log"
+	"os"
+	"os/exec"
+	"syscall"
+)
+
+// CgroupImpl breaks out the cgroup functionality so the tests don't need elevated privileges.
+type CgroupImpl interface {
+	MakeCgroup(cg *ControlGroup, ppid Process) error
+	Run(ctx context.Context, wrapper string, spa *syscall.SysProcAttr, out *SafeBuffer, cleanup func(), cgDir string, c string, args ...string) (*exec.Cmd, *StatusTracker, <-chan bool, error)
+}
+
+type RealCgroup struct{}
+
+func (r *RealCgroup) MakeCgroup(cg *ControlGroup, ppid Process) error {
+	// makes cgroups in the v2 style of /sys/fs/cgroup/<ppid>
+	group := fmt.Sprintf("/sys/fs/cgroup/%s", ppid)
+	if err := os.MkdirAll(group, 0644); err != nil {
+		if errors.Is(err, fs.ErrExist) {
+			log.Println("cgroup", group, "already exists")
+		} else {
+			log.Println("failed to create cgroup for", group, ":", err)
+			return err
+		}
+	}
+
+	// Now we can write the actual cgroup limits ...
+	for _, l := range cg.Limits {
+		lf := fmt.Sprintf("%s/%s", group, l.Var)
+		setVar(l.Value, lf)
+	}
+	return nil
+}
+
+func (r *RealCgroup) Run(ctx context.Context, wrapper string, spa *syscall.SysProcAttr, out *SafeBuffer, cleanup func(), cgDir string, c string, args ...string) (*exec.Cmd, *StatusTracker, <-chan bool, error) {
+	args = append([]string{cgDir, c}, args...)
+	log.Println("running process:", wrapper, c, args)
+	cmd := exec.Command(wrapper, args...)
+
+	cmd.Stdout = out
+	cmd.Stderr = out
+
+	// NOTE: experimental - we may need to do this in the wrapper.
+	cmd.SysProcAttr = spa // linux namespace controls
+
+	done := make(chan bool)
+	tracker := &StatusTracker{Status: Unavailable}
+
+	if err := cmd.Start(); err != nil {
+		fmt.Printf("Error running the exec.Command - %s\n", err)
+		cleanup()
+		os.Exit(1)
+	}
+
+	// NOTE: wrapper will add pid to procs file
+
+	go func() {
+		tracker.SetStatus(Running, 0)
+		werr := cmd.Wait()
+		if werr != nil {
+			var ec int
+			if ee, ok := werr.(*exec.ExitError); !ok {
+				ec = ee.ExitCode()
+			}
+			tracker.SetStatus(Error, ec)
+		} else {
+			tracker.SetStatus(Completed, 0)
+		}
+		log.Println("... process ended")
+		done <- true
+		cleanup()
+	}()
+
+	return cmd, tracker, done, nil
+}
+
+func alterController(cmd, fn string) {
+	writeWithFlags(cmd, fn, os.O_WRONLY|os.O_APPEND)
+}
+
+func setVar(cmd, fn string) {
+	writeWithFlags(cmd, fn, os.O_CREATE|os.O_WRONLY)
+}
```

Comment:
I don't think you need O_APPEND or O_CREATE|O_WRONLY to manage cgroups; pretty sure that both of these can be replaced with `os.WriteFile`.